dev/Tiles/entity.h

```diff
@@ -0,0 +1,66 @@
+#include <stdlib.h>
+#include <stdbool.h>
+#include "metatile.h"
+
+typedef struct Entity {
+    MetaTile* metatile[32];
```

Comment:
You do things like this quite a lot, and it's a little confusing to me; I'm not totally sure you are doing what you intend to.

This declares an array of 32 MetaTile pointers, so you can have up to 32 MetaTiles in your Entity, although it looks like from your implementation that you don't enforce that limit anywhere, so it could be very easy to accidentally run off the end of your array.

My suggestion would either be to use an unbounded array (pointer) type, e.g. `MetaTile *metatile[]` or `MetaTile **metatile[]` and a length parameter, or else a linked-list structure within MetaTile itself.

Reply:
Could you describe the double pointer mechanism with a length a little further please. It looks like something I've seen before.

Reply:
Yep, I'll try and work with what I think you're trying to do and use that as a context.

You have a structure, `Entity`, which you intend to reference a variable number of MetaTiles. You are not attempting to store the MetaTiles in the structure itself, so therefore you must be storing pointers.

It also seems like you are using the `numTiles` field to hold the number of MetaTiles you are storing, so I think we also do have a length parameter.

Now the question is, what type do we use to store a variable length array within a struct? We can't use
`MetaTile metatile[]` because you would need to supply a fixed length for that, and in any case, if you did supply a fixed length the compiler would then make the struct big enough to hold all those elements, which is not what you want at all. Changing it to `MetaTile *metatile[]` is no better, because C still doesn't like that you haven't supplied a length for the array, and doing so will still cause the struct to allocate a fixed amount of space, although now the amount of space allocated will (probably) be smaller because you're storing a fixed number of pointers, rather than a fixed number of structures.

What you really want is to be able to write `MetaTile *metatile[]` somehow and have C be okay with it.
You have two choices:

1) Write `MetaTile *metatile` and make sure all your MetaTiles are stored consecutively, e.g. in an actual array somewhere. Then use pointer arithmetic to navigate around them and use your `numTiles` field to detect whether you are pointing at something valid:

```
void print_tiles(Entity *e)
{
  for (int i = 0; i < e->num_tiles; i++)
  {
    printf("%d\n", e->metatiles[i].some_field);
  }
}

// static allocation, e.g. 3 tiles
MetaTile tiles[3];
Entity e = { .num_tiles = 3, .metatiles = tiles };
for (int i = 0; i < e.num_tiles; i++)
{
  e.metatiles[i].some_field = 5;
}
print_entity(&e);
```
or
```
// dynamic allocation, e.g. 3 tiles
MetaTile *tiles = malloc(3*sizeof(MetaTile));
Entity e = { .num_tiles = 3, .metatiles = tiles };
for (int i = 0; i < e.num_tiles; i++)
{
  e.metatiles[i].some_field = 3;
}
print_entity(&e);
```

In this case, your `metatile` field really only contains a pointer to the _first_ MetaTile in a consecutive set of them (an array), and the relationship between the first and second and subsequent tiles is implied only.

2) Write `MetaTile **metatile`. In other words, you are creating a pointer to a pointer to a Metatile. Why would you want to do this? Because you can satisfy the first level of pointing with a simple array. In other words, somewhere in memory you have an array of pointers to MetaTiles, and you are passing around a pointer to that array in your structure:

```
// static allocation, e.g. 3 tiles
MetaTile *tiles[3];
Entity e = { .num_tiles = 3, .metatiles = tiles };
for (int i = 0; i < e.num_tiles; i++)
{
  MetaTile t = { .some_field = 5 };
  e.metatiles[i] = &t;
}
print_tiles(&e);
```

or

```
// dynamic allocation, e.g. 3 tiles
MetaTile **tiles = malloc(3*sizeof(MetaTile*));
Entity e = { .num_tiles = 3, .metatiles = tiles };
for (int i = 0; i < e.num_tiles; i++)
{
  MetaTile *t = malloc(sizeof(MetaTile));
  e.metatiles[i] = t;
  e.metatiles[i]->some_field = 3;
}
print_tiles(&e);
```